_blog/2023-12-15-using-actix-rust.mdx

```diff
@@ -0,0 +1,434 @@
+    type Future = Ready<Result<Self::Transform, Self::InitError>>;
+
+    // this immediately returns the middleware
+    fn new_transform(&self, service: S) -> Self::Future {
+        ready(Ok(SayHiMiddleware { service }))
+    }
+}
+```
+Now we can write the middleware itself! Internally the middleware must implement a generic type - which then gets declared in the `Service` trait. Note that we manually re-implement a type from `futures_util` called `LocalBoxFuture` - that is to say, a future that doesn't require the `Send` trait and is safe to use because it implements `Unpin` on dereferencing, which automatically cancels any previous thread-safety guarantees.
+```rust
+pub struct SayHiMiddleware<S> {
+    /// The next service to call
+    service: S,
+}
+
+// This future doesn't have the requirement of being `Send`.
+// See: futures_util::future::LocalBoxFuture
+type LocalBoxFuture<T> = Pin<Box<dyn Future<Output = T> + 'static>>;
+
+// `S`: type of the wrapped service
+// `B`: type of the body - try to be generic over the body where possible
+impl<S, B> Service<ServiceRequest> for SayHiMiddleware<S>
+where
+    S: Service<ServiceRequest, Response = ServiceResponse<B>, Error = Error>,
+    S::Future: 'static,
+    B: 'static,
+{
+    type Response = ServiceResponse<B>;
+    type Error = Error;
+    type Future = LocalBoxFuture<Result<Self::Response, Self::Error>>;
+
+    // This service is ready when its next service is ready
+    forward_ready!(service);
+
+    fn call(&self, req: ServiceRequest) -> Self::Future {
+        println!("Hi from start. You requested: {}", req.path());
+
+        // A more complex middleware, could return an error or an early response here.
+
+        // we do not immediately await this, which means nothing happens
+        // this future gets moved into a Box
+        let fut = self.service.call(req);
+
+        Box::pin(async move {
+            // this future gets awaited now
+            let res = fut.await?;
+
+            // we can now do any work we need to after the request
+            println!("Hi from response");
+            Ok(res)
+        })
+    }
+}
+```
+Now that we've written our middleware, we can now add it to our App:
+```rust
+#[actix_web::main]
+async fn main() -> std::io::Result<()> {
+// access logs are printed with the INFO level so ensure it is enabled by default
+env_logger::init_from_env(env_logger::Env::new().default_filter_or("info"));
+
+let app = App::new()
+     .wrap(SayHi);
+     // ... rest of your code
+}
+```
+
+## Static Files in Actix Web
+Plain, no-frills static file serving in [`actix-web`](https://actix.rs/) is done through the `actix_files` crate - to add it, you simply need to add it through Cargo like so:
+```bash
+cargo add actix-files 
+```
+
+Setting up a route for static file serving would look like this:
+```rust
+use actix_files::NamedFile;
+use actix_web::{HttpRequest, Result};
+use std::path::PathBuf;
+
+async fn index(req: HttpRequest) -> Result<NamedFile> {
```

Comment:
Is this one missing the macro?

Reply:
Just went through the article and added the macro to all of the routes